discovery/src/main/java/com/alibaba/fescar/discovery/registry/consul/ConsulRegistryServiceImpl.java

```diff
@@ -0,0 +1,327 @@
+    }
+
+    /**
+     * create service check based on TCP
+     *
+     * @param address
+     * @return
+     */
+    private NewService.Check createCheck(InetSocketAddress address) {
+        NewService.Check check = new NewService.Check();
+        check.setTcp(NetUtil.toStringAddress(address));
+        check.setInterval(DEFAULT_CHECK_INTERVAL);
+        check.setTimeout(DEFAULT_CHECK_TIMEOUT);
+        check.setDeregisterCriticalServiceAfter(DEFAULT_DEREGISTER_TIME);
+        return check;
+    }
+
+    /**
+     * get healthy services
+     *
+     * @param service
+     * @return
+     */
+    private Response<List<HealthService>> getHealthyServices(String service, long index, long watchTimeout) {
+        return getConsulClient().getHealthServices(service, HealthServicesRequest.newBuilder()
+            .setTag(SERVICE_TAG)
+            .setQueryParams(new QueryParams(watchTimeout, index))
+            .setPassing(true)
+            .build());
+    }
+
+    /**
+     * get service group
+     *
+     * @param key
+     * @return clusterNameKey
+     */
+    private String getServiceGroup(String key) {
+        Configuration configuration = ConfigurationFactory.getInstance();
+        String clusterNameKey = PREFIX_SERVICE_ROOT + CONFIG_SPLIT_CHAR + PREFIX_SERVICE_MAPPING + key;
+        return configuration.getConfig(clusterNameKey);
+    }
+
+    /**
+     * refresh cluster
+     *
+     * @param cluster
+     */
+    private void refreshCluster(String cluster) {
+        if (null == cluster) {
+            return;
+        }
+        Response<List<HealthService>> response = getHealthyServices(getClusterName(), -1, -1);
+        if (response == null) {
+            return;
+        }
+        refreshCluster(cluster, response.getValue());
+    }
+
+    /**
+     * refresh cluster
+     *
+     * @param cluster
+     * @param services
+     */
+    private void refreshCluster(String cluster, List<HealthService> services) {
+        if (null == cluster || services == null) {
+            return;
+        }
+        clusterAddressMap.put(cluster, services.stream()
+            .map(HealthService::getService)
+            .map(service -> new InetSocketAddress(service.getAddress(), service.getPort()))
+            .collect(Collectors.toList()));
+    }
+
+    /**
+     * consul notifier
+     */
+    private class ConsulNotifier implements Runnable {
+        private String cluster;
+        private long consulIndex;
+        private boolean running;
+
+        ConsulNotifier(String cluster, long consulIndex) {
+            this.cluster = cluster;
+            this.consulIndex = consulIndex;
+            this.running = true;
+        }
+
+        @Override
+        public void run() {
+            while (this.running) {
+                processService();
+            }
+        }
+
+        private void processService() {
+            Response<List<HealthService>> response = getHealthyServices(cluster, consulIndex, DEFAULT_WATCH_TIMEOUT);
```

Comment:
will hang？

Reply:
yes.

Reply:
In consul, all of the operations are web-based and without an event listener provided In the API. However, it provides a Blocking query to monitor a change in the state of an instance. The specific process is that the service instance is queried first, and an consul index will be included in the returned result. Then, make new query with the returned consul index.and the query will be hung until the status of the instance changes, and the new consul index will be returned and it will be larger than the previous one.